File: iOS/Objc/Objc/GenerateKeystore.m
//
//  ViewController.m
//  Objc
//
//  Created by Shilei Mao on 03/08/2021.
//

#import "GenerateKeystore.h"
#import <HorizonFintexSDK/HorizonFintexSDK.h>

@interface GenerateKeystore ()
@property(strong, nonatomic) HGEthereumOperator *ethOperator;
@end

@implementation GenerateKeystore

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    JSWebview *webview = [[JSWebview alloc] initWithHandlers: nil];
    
    self.ethOperator = [[HGEthereumOperator alloc] initWithJsEvaluator:webview attachedView:self.view];
    
}

-(IBAction) btGenerateKeystoreTapped:(id)sender {
    NSString *pwdStr = self.tfPassword.text;
    if (pwdStr == nil || [pwdStr length] == 0) {
        [self showErrorWithMessage:@"Please enter password"];
        return;
    }
    
    [self showProgress];
    [self.ethOperator generateKeystoreWithPwd:pwdStr completion:^(EthereumKeystore * _Nullable keystore, NSError * _Nullable error) {
        [self dismissProgress];
        self.tvKeystore.text = [keystore keystoreStr];
        NSLog(@"result: %@", keystore);
    }];
}


- (void)showErrorWithMessage:(NSString *)message {
    UIAlertController *alertView = [UIAlertController
                                    alertControllerWithTitle:@"Error" message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *cancelAction = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleCancel handler:nil];
    
    [alertView addAction:cancelAction];
    
    [self presentViewController:alertView animated:true completion:nil];
}

- (void)showProgress {
    UIViewController *progressView = [self.storyboard instantiateViewControllerWithIdentifier:@"ProgressView"];
    [progressView setModalTransitionStyle:UIModalTransitionStyleCrossDissolve];
    [progressView setModalPresentationStyle:UIModalPresentationOverFullScreen];
    
    [self presentViewController:progressView animated:true completion:nil];
}

- (void)dismissProgress {
    [self dismissViewControllerAnimated:true completion:nil];
}

@end

File: iOS/Swift/SwiftTest/GenerateKeystore.swift
//
//  ViewController.swift
//  Swift
//
//  Created by Shilei Mao on 03/08/2021.
//

import UIKit
import HorizonFintexSDK

class GenerateKeystore: UIViewController {

    @IBOutlet weak var tfGenerateKeystorePwd: UITextField!
    
    @IBOutlet weak var btGenerateKeystore: UIButton!
    
    @IBOutlet weak var tvKeystore: UITextView!
    
    
    var ethOperator: HGEthereumOperator!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        let jsEvaluator = JSWebview(handlers: nil)
        self.ethOperator = HGEthereumOperator(jsEvaluator: jsEvaluator, attachedView: self.view)
    }
    
    @IBAction func btGeneratedKeystoreTapped(_ sender: Any) {
        if let pwd = self.tfGenerateKeystorePwd.text, !pwd.isEmpty {
            self.showProgress()
            self.ethOperator.generateKeystore(pwd: pwd) { keystore, error in
                self.tvKeystore.text = keystore?.keystoreStr
                self.dismissProgress()
            }
        } else {
            showError(message: "Please enter password")
        }
    }
    
    
    func showError(message: String) {
        let alert = UIAlertController(title: "Error", message: message, preferredStyle: .alert)
        let cancelAction = UIAlertAction(title: "Ok", style: .cancel, handler: nil)
        
        alert.addAction(cancelAction)
        self.present(alert, animated: true, completion: nil)
    }
    
    func showProgress() {
        if let progressView = storyboard?.instantiateViewController(identifier: "ProgressView") {
            progressView.modalTransitionStyle = .crossDissolve
            progressView.modalPresentationStyle = .overCurrentContext
            
            self.present(progressView, animated: true, completion: nil)
        }
    }
    
    func dismissProgress() {
        self.dismiss(animated: true, completion: nil)
    }
}

